### ISSUE TITLE
readVcf is Slow if ScanVcfParam which Regions is Lengthy

### ISSUE THREAD

**Issue by DarioS** [Bioconductor/VariantAnnotation#70]
Recently, U.S. Food and Drug Administration, National Institute of Standards and Technology and Illumina [researchers defined highly reproducible regions (H.R.R.s)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02569-8) of the human genome and [made available BED files](https://zenodo.org/record/5275189#.ZCDkyMpBzIX) to define a set of regions in which variant callers consistently call variants on technical replicate samples, effectively defining a while-list for whole genome sequencing data. `readVcf` takes a long time if `which` of `ScanVcfParam` is specified. Importing the V.C.F. takes about five minutes if `which` not specified but I terminated it after one hour when `which` was specified.

```r
library(rtracklayer)
goldStandards <- list.files("HRR/", "bed", full.names = TRUE)
goldStandards <- lapply(goldStandards, import.bed)
goldStandards <- unlist(goldStandards)
goldStandards <- reduce(goldStandards)
> goldStandards
GRanges object with 2988875 ranges and 0 metadata columns:
    ...        ...
variants <- readVcf("DRAGENgermline.vcf.gz", param = ScanVcfParam(which = goldStandards)) # Stopped after one hour.
```
Importing the whole V.C.F. file into the session and then using `subsetByOverlaps` seems a reasonable and fast workaround. To best help other users, would a documentation change or code change be better to make this user experience nicer?

**Comment by vjcitn** [comment]
Interesting observation @DarioS .  This topic (using the HRR resource) could merit a workflow package, which could then be referenced by VariantAnnotation documentation.  Would you be up for this?  A public VCF could be used to demonstrate the approach.  Or a PR to the VariantAnnotation doc could be a rapid resolution of your concern.  Did you consider breaking up goldStandards and performing the read on sorted, modest-sized, single-chromosome chunks?

**Comment by mtmorgan** [comment]
Does the [filterVcf](https://bioconductor.org/packages/release/bioc/vignettes/VariantAnnotation/inst/doc/filterVcf.pdf) vignette address this use case?

**Comment by DarioS** [comment]
Indeed, Section 5.2 of filterVcf cignette recommends `readVcf` followed by `findOverlaps`. But Section 7 Appendix seems to advocate for using `ScanVcfParam(which = theRegions)`. It's unclear which approach is better and for which scenarios.

**Comment by jmg421** [comment]
This is a known limitation of how tabix works: htslib makes one random-access seek per range in `which(param)`. With ~3M ranges (as in the HRR BED file) that means ~3M seeks, which is far slower than streaming the whole file once.

**Recommendations:**
1. **Merge nearby ranges first**: `readVcf(file, param = ScanVcfParam(which = reduce(goldStandards, min.gapwidth=500L)))` — merging ranges that are within 500bp reduces the seek count dramatically with minimal loss of precision.
2. **Stream then filter**: as you found, reading the whole VCF and using `subsetByOverlaps` is faster when the `which` list is very long.

I'll add a note to the `ScanVcfParam` docs warning that very long `which` vectors incur one tabix seek per range and recommending `reduce()` as a pre-processing step.
